File: backend/app/services/report_service.py
"""Reporting service - SQL aggregation queries for the dashboard and reports."""
from __future__ import annotations

import csv
import io
from typing import Dict, List

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.models.approval import Approval, ApprovalDecision
from app.models.task import Task, TaskPriority, TaskStatus
from app.models.user import User
from app.schemas.report import (
    ApprovalSummaryEntry,
    ApprovalSummaryReport,
    PriorityCount,
    StatusCount,
    TasksByPriorityReport,
    TasksByStatusReport,
    UserWorkloadEntry,
    UserWorkloadReport,
)


class ReportService:
    """Aggregation queries grouped by status, priority, user, and decision."""

    def __init__(self, db: Session) -> None:
        self.db = db

    def tasks_by_status(self) -> TasksByStatusReport:
        stmt = select(Task.status, func.count(Task.id)).group_by(Task.status)
        rows = self.db.execute(stmt).all()
        counts: Dict[TaskStatus, int] = {s: 0 for s in TaskStatus}
        for status, count in rows:
            counts[status] = int(count)
        items = [StatusCount(status=key, count=value) for key, value in counts.items()]
        return TasksByStatusReport(items=items, total=sum(counts.values()))

    def tasks_by_priority(self) -> TasksByPriorityReport:
        stmt = select(Task.priority, func.count(Task.id)).group_by(Task.priority)
        rows = self.db.execute(stmt).all()
        counts: Dict[TaskPriority, int] = {p: 0 for p in TaskPriority}
        for priority, count in rows:
            counts[priority] = int(count)
        items = [PriorityCount(priority=key, count=value) for key, value in counts.items()]
        return TasksByPriorityReport(items=items, total=sum(counts.values()))

    def user_workload(self) -> UserWorkloadReport:
        users_stmt = select(User).order_by(User.name)
        users = list(self.db.scalars(users_stmt).all())

        breakdown_stmt = (
            select(Task.assigned_to_user_id, Task.status, func.count(Task.id))
            .group_by(Task.assigned_to_user_id, Task.status)
        )
        rows = self.db.execute(breakdown_stmt).all()

        per_user: Dict[int, Dict[TaskStatus, int]] = {
            user.id: {s: 0 for s in TaskStatus} for user in users
        }
        for user_id, status, count in rows:
            if user_id in per_user:
                per_user[user_id][status] = int(count)

        items: List[UserWorkloadEntry] = []
        for user in users:
            counts = per_user[user.id]
            items.append(
                UserWorkloadEntry(
                    user_id=user.id,
                    name=user.name,
                    email=user.email,
                    role=user.role,
                    total=sum(counts.values()),
                    pending=counts[TaskStatus.PENDING],
                    in_progress=counts[TaskStatus.IN_PROGRESS],
                    submitted=counts[TaskStatus.SUBMITTED],
                    approved=counts[TaskStatus.APPROVED],
                    rejected=counts[TaskStatus.REJECTED],
                    completed=counts[TaskStatus.COMPLETED],
                )
            )
        return UserWorkloadReport(items=items)

    def approval_summary(self) -> ApprovalSummaryReport:
        stmt = (
            select(Approval.decision, func.count(Approval.id))
            .group_by(Approval.decision)
        )
        rows = self.db.execute(stmt).all()
        counts: Dict[ApprovalDecision, int] = {d: 0 for d in ApprovalDecision}
        for decision, count in rows:
            counts[decision] = int(count)
        items = [
            ApprovalSummaryEntry(decision=key, count=value) for key, value in counts.items()
        ]
        return ApprovalSummaryReport(items=items, total=sum(counts.values()))

    def export_tasks_csv(self) -> str:
        """Build a CSV string containing all tasks for download."""
        stmt = (
            select(Task, User.name.label("assignee_name"), User.email.label("assignee_email"))
            .join(User, User.id == Task.assigned_to_user_id)
            .order_by(Task.id)
        )
        rows = self.db.execute(stmt).all()

        buffer = io.StringIO()
        writer = csv.writer(buffer)
        writer.writerow(
            [
                "id",
                "title",
                "description",
                "status",
                "priority",
                "assignee_name",
                "assignee_email",
                "created_by_user_id",
                "created_at",
                "updated_at",
            ]
        )
        for task, assignee_name, assignee_email in rows:
            writer.writerow(
                [
                    task.id,
                    task.title,
                    task.description.replace("\n", " "),
                    task.status.value,
                    task.priority.value,
                    assignee_name,
                    assignee_email,
                    task.created_by_user_id,
                    task.created_at.isoformat() if task.created_at else "",
                    task.updated_at.isoformat() if task.updated_at else "",
                ]
            )
        return buffer.getvalue()

File: backend/src/main/java/com/example/bwms/service/ReportService.java
// Translated from: backend/app/services/report_service.py
// NOTE: user_workload merges two queries in Java memory — same approach as the Python original.
//       See Change Log for discussion of a more efficient single-query alternative.
package com.example.bwms.service;

import com.example.bwms.dto.report.*;
import com.example.bwms.model.*;
import com.example.bwms.repository.ApprovalRepository;
import com.example.bwms.repository.TaskRepository;
import com.example.bwms.repository.UserRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.*;

@Service
public class ReportService {

    private final TaskRepository taskRepository;
    private final UserRepository userRepository;
    private final ApprovalRepository approvalRepository;

    public ReportService(TaskRepository taskRepository, UserRepository userRepository,
                         ApprovalRepository approvalRepository) {
        this.taskRepository = taskRepository;
        this.userRepository = userRepository;
        this.approvalRepository = approvalRepository;
    }

    @Transactional(readOnly = true)
    public TasksByStatusReport tasksByStatus() {
        Map<TaskStatus, Integer> counts = new EnumMap<>(TaskStatus.class);
        for (TaskStatus s : TaskStatus.values()) counts.put(s, 0);

        for (Object[] row : taskRepository.countByStatus()) {
            TaskStatus status = (TaskStatus) row[0];
            counts.put(status, ((Long) row[1]).intValue());
        }

        List<StatusCountDto> items = new ArrayList<>();
        for (Map.Entry<TaskStatus, Integer> e : counts.entrySet()) {
            items.add(new StatusCountDto(e.getKey(), e.getValue()));
        }
        int total = counts.values().stream().mapToInt(Integer::intValue).sum();
        return new TasksByStatusReport(items, total);
    }

    @Transactional(readOnly = true)
    public TasksByPriorityReport tasksByPriority() {
        Map<TaskPriority, Integer> counts = new EnumMap<>(TaskPriority.class);
        for (TaskPriority p : TaskPriority.values()) counts.put(p, 0);

        for (Object[] row : taskRepository.countByPriority()) {
            TaskPriority priority = (TaskPriority) row[0];
            counts.put(priority, ((Long) row[1]).intValue());
        }

        List<PriorityCountDto> items = new ArrayList<>();
        for (Map.Entry<TaskPriority, Integer> e : counts.entrySet()) {
            items.add(new PriorityCountDto(e.getKey(), e.getValue()));
        }
        int total = counts.values().stream().mapToInt(Integer::intValue).sum();
        return new TasksByPriorityReport(items, total);
    }

    @Transactional(readOnly = true)
    public UserWorkloadReport userWorkload() {
        List<User> users = userRepository.findAllByOrderByNameAsc();

        // Initialise zero-count map for every user × status combination
        Map<Long, Map<TaskStatus, Integer>> perUser = new LinkedHashMap<>();
        for (User user : users) {
            Map<TaskStatus, Integer> statusMap = new EnumMap<>(TaskStatus.class);
            for (TaskStatus s : TaskStatus.values()) statusMap.put(s, 0);
            perUser.put(user.getId(), statusMap);
        }

        for (Object[] row : taskRepository.countByAssigneeAndStatus()) {
            Long userId = (Long) row[0];
            TaskStatus status = (TaskStatus) row[1];
            int count = ((Long) row[2]).intValue();
            if (perUser.containsKey(userId)) {
                perUser.get(userId).put(status, count);
            }
        }

        List<UserWorkloadEntry> items = new ArrayList<>();
        for (User user : users) {
            Map<TaskStatus, Integer> c = perUser.get(user.getId());
            int total = c.values().stream().mapToInt(Integer::intValue).sum();
            items.add(new UserWorkloadEntry(
                    user.getId(), user.getName(), user.getEmail(), user.getRole(),
                    total,
                    c.get(TaskStatus.PENDING),
                    c.get(TaskStatus.IN_PROGRESS),
                    c.get(TaskStatus.SUBMITTED),
                    c.get(TaskStatus.APPROVED),
                    c.get(TaskStatus.REJECTED),
                    c.get(TaskStatus.COMPLETED)
            ));
        }
        return new UserWorkloadReport(items);
    }

    @Transactional(readOnly = true)
    public ApprovalSummaryReport approvalSummary() {
        Map<ApprovalDecision, Integer> counts = new EnumMap<>(ApprovalDecision.class);
        for (ApprovalDecision d : ApprovalDecision.values()) counts.put(d, 0);

        for (Object[] row : approvalRepository.countByDecision()) {
            ApprovalDecision decision = (ApprovalDecision) row[0];
            counts.put(decision, ((Long) row[1]).intValue());
        }

        List<ApprovalSummaryEntry> items = new ArrayList<>();
        for (Map.Entry<ApprovalDecision, Integer> e : counts.entrySet()) {
            items.add(new ApprovalSummaryEntry(e.getKey(), e.getValue()));
        }
        int total = counts.values().stream().mapToInt(Integer::intValue).sum();
        return new ApprovalSummaryReport(items, total);
    }

    @Transactional(readOnly = true)
    public String exportTasksCsv() {
        List<Task> tasks = taskRepository.findAllWithAssigneeOrderById();
        StringBuilder sb = new StringBuilder();
        sb.append("id,title,description,status,priority,assignee_name,assignee_email,created_by_user_id,created_at,updated_at\n");
        for (Task task : tasks) {
            User assignee = task.getAssignee();
            sb.append(csvRow(
                    String.valueOf(task.getId()),
                    task.getTitle(),
                    task.getDescription().replace("\n", " "),
                    task.getStatus().getValue(),
                    task.getPriority().getValue(),
                    assignee.getName(),
                    assignee.getEmail(),
                    String.valueOf(task.getCreatedByUserId()),
                    task.getCreatedAt() != null ? task.getCreatedAt().toString() : "",
                    task.getUpdatedAt() != null ? task.getUpdatedAt().toString() : ""
            ));
        }
        return sb.toString();
    }

    private String csvRow(String... fields) {
        StringBuilder row = new StringBuilder();
        for (int i = 0; i < fields.length; i++) {
            if (i > 0) row.append(',');
            row.append(csvEscape(fields[i]));
        }
        row.append('\n');
        return row.toString();
    }

    private String csvEscape(String value) {
        if (value == null) return "";
        if (value.contains(",") || value.contains("\"") || value.contains("\n")) {
            return "\"" + value.replace("\"", "\"\"") + "\"";
        }
        return value;
    }
}
